File: packages/python/src/blackrose/decide.py
"""Map TypeSafe System One answers onto CheckResult verdicts."""

from __future__ import annotations

import math
from collections.abc import Iterable, Mapping
from typing import Any

from blackrose.policy import Policy
from blackrose.types import CheckResult, Trigger, Verdict

_PRECEDENCE: tuple[Verdict, ...] = ("block", "review", "allow")


def _attr(obj: Any, name: str, default: Any = None) -> Any:
    if obj is None:
        return default
    if isinstance(obj, Mapping):
        return obj.get(name, default)
    return getattr(obj, name, default)


def _has_field(obj: Any, name: str) -> bool:
    if obj is None:
        return False
    if isinstance(obj, Mapping):
        return name in obj
    return hasattr(obj, name)


def _as_float(value: Any) -> float | None:
    if value is None or isinstance(value, bool):
        return None
    if isinstance(value, str) and value.strip() == "":
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(number):
        return None
    return number


def _is_answer_like(value: Any) -> bool:
    if value is None or isinstance(value, (str, int, float, bool)):
        return False
    return _has_field(value, "noul") or _has_field(value, "score") or _has_field(value, "choice")


def answers_view(raw: Any) -> Mapping[str, Any]:
    """Normalize SDK response / mock into a name → answer mapping."""
    if raw is None:
        return {}
    if _has_field(raw, "answers"):
        answers = _attr(raw, "answers")
        return answers if isinstance(answers, Mapping) else {}

    combined: dict[str, Any] = {}
    for bucket in ("nouls", "scores", "choices"):
        group = _attr(raw, bucket)
        if isinstance(group, Mapping):
            combined.update(group)
    if combined:
        return combined

    if isinstance(raw, Mapping):
        filtered = {name: value for name, value in raw.items() if _is_answer_like(value)}
        return filtered
    return {}


def extract_scores(raw: Any) -> dict[str, float]:
    """Pull named numeric signals from a TypeSafe response or mock dict."""
    scores: dict[str, float] = {}
    for name, answer in answers_view(raw).items():
        noul = _as_float(_attr(answer, "noul"))
        if noul is not None:
            scores[name] = noul
            continue
        score = _as_float(_attr(answer, "score"))
        if score is not None:
            scores[name] = score
            continue
        choice = _attr(answer, "choice")
        probs = _attr(answer, "probabilities")
        if choice is not None and isinstance(probs, Mapping):
            p = _as_float(probs.get(choice))
            if p is not None:
                scores[name] = p
    return scores


def _qualified(trigger: Trigger) -> str:
    return trigger.qualified()


def _low_confidence_trigger(
    name: str,
    confidence: float | None,
    *,
    missing: bool,
    min_confidence: float,
) -> Trigger:
    if missing:
        message = f"{name} confidence missing (< min_confidence {min_confidence})"
    else:
        value = 0.0 if confidence is None else confidence
        message = f"{name} confidence={value:.2f} < min_confidence ({min_confidence})"
    return Trigger(code="low_confidence", check=name, message=message)


def decide(
    raw: Any,
    policy: Policy,
    *,
    expected_checks: Iterable[str] | None = None,
) -> CheckResult:
    """Apply policy thresholds to TypeSafe answers.

    Precedence: any ``block`` wins over ``review``; otherwise ``allow``.
    Missing expected checks, empty responses, and low/missing Score/Choice
    confidence contribute ``review`` (never silent ``allow``).
    """
    answers = answers_view(raw)
    scores = extract_scores(raw)
    triggers: list[Trigger] = []
    triggered: set[Verdict] = set()

    expected = list(expected_checks) if expected_checks is not None else None
    if expected is not None:
        for name in expected:
            if name not in answers:
                triggered.add("review")
                triggers.append(
                    Trigger(
                        code="missing_check",
                        check=name,
                        message=f"{name} missing from TypeSafe response",
                    )
                )
        if not expected and not answers:
            triggered.add("review")
            triggers.append(
                Trigger(
                    code="empty_response",
                    check=None,
                    message="TypeSafe response contained no answers",
                )
            )
    elif not answers:
        triggered.add("review")
        triggers.append(
            Trigger(
                code="empty_response",
                check=None,
                message="TypeSafe response contained no answers",
            )
        )

    for name, answer in answers.items():
        noul = _as_float(_attr(answer, "noul")) if _has_field(answer, "noul") else None
        if _has_field(answer, "noul"):
            if noul is not None:
                if noul >= policy.block_threshold and name in policy.block_checks:
                    triggered.add("block")
                    triggers.append(
                        Trigger(
                            code="noul_block",
                            check=name,
                            message=(
                                f"{name}={noul:.2f} >= block_threshold ({policy.block_threshold})"
                            ),
                        )
                    )
                elif noul >= policy.review_threshold:
                    triggered.add("review")
                    triggers.append(
                        Trigger(
                            code="noul_review",
                            check=name,
                            message=(
                                f"{name}={noul:.2f} >= review_threshold ({policy.review_threshold})"
                            ),
                        )
                    )
            if _has_field(answer, "confidence"):
                confidence = _as_float(_attr(answer, "confidence"))
                if confidence is not None and confidence < policy.min_confidence:
                    triggered.add("review")
                    triggers.append(
                        _low_confidence_trigger(
                            name,
                            confidence,
                            missing=False,
                            min_confidence=policy.min_confidence,
                        )
                    )
            continue

        if _has_field(answer, "score"):
            score = _as_float(_attr(answer, "score"))
            confidence = _as_float(_attr(answer, "confidence"))
            missing_confidence = not _has_field(answer, "confidence") or confidence is None
            low_confidence = confidence is not None and confidence < policy.min_confidence
            if missing_confidence or low_confidence:
                triggered.add("review")
                triggers.append(
                    _low_confidence_trigger(
                        name,
                        confidence,
                        missing=missing_confidence,
                        min_confidence=policy.min_confidence,
                    )
                )
            cutoffs = policy.score_cutoffs(name)
            if score is not None and cutoffs is not None:
                if score >= cutoffs.block:
                    triggered.add("block")
                    triggers.append(
                        Trigger(
                            code="score_block",
                            check=name,
                            message=f"{name}={score:.2f} >= score_block ({cutoffs.block})",
                        )
                    )
                elif score >= cutoffs.review:
                    triggered.add("review")
                    triggers.append(
                        Trigger(
                            code="score_review",
                            check=name,
                            message=f"{name}={score:.2f} >= score_review ({cutoffs.review})",
                        )
                    )
            continue

        # Choice answers: low or missing confidence → review; no default block path.
        if _has_field(answer, "choice") or _has_field(answer, "confidence"):
            confidence = _as_float(_attr(answer, "confidence"))
            missing_confidence = not _has_field(answer, "confidence") or confidence is None
            low_confidence = confidence is not None and confidence < policy.min_confidence
            if missing_confidence or low_confidence:
                triggered.add("review")
                triggers.append(
                    _low_confidence_trigger(
                        name,
                        confidence,
                        missing=missing_confidence,
                        min_confidence=policy.min_confidence,
                    )
                )

    verdict: Verdict = next((v for v in _PRECEDENCE if v in triggered), "allow")
    return CheckResult(
        verdict=verdict,
        reasons=[t.message for t in triggers],
        scores=scores,
        raw=raw,
        triggers=tuple(triggers),
        codes=tuple(_qualified(t) for t in triggers),
    )

File: packages/js/src/decide.ts
/**
 * Map TypeSafe System One answers onto CheckResult verdicts.
 *
 * Mirrors packages/python/src/blackrose/decide.py.
 */

import type { Policy } from "./policy.js";
import type { CheckResult, Trigger, Verdict } from "./types.js";
import { qualifiedTrigger } from "./types.js";

const PRECEDENCE: readonly Verdict[] = ["block", "review", "allow"];

function hasField(obj: unknown, name: string): boolean {
  if (obj == null || typeof obj !== "object") return false;
  return name in (obj as Record<string, unknown>);
}

function attr(obj: unknown, name: string, fallback: unknown = undefined): unknown {
  if (!hasField(obj, name)) return fallback;
  return (obj as Record<string, unknown>)[name];
}

function asFloat(value: unknown): number | null {
  if (value == null || typeof value === "boolean") return null;
  if (typeof value === "string") {
    const trimmed = value.trim();
    if (trimmed === "") return null;
    const n = Number(trimmed);
    return Number.isFinite(n) ? n : null;
  }
  if (typeof value === "number") {
    return Number.isFinite(value) ? value : null;
  }
  return null;
}

function isAnswerLike(value: unknown): boolean {
  if (value == null || typeof value !== "object") return false;
  return hasField(value, "noul") || hasField(value, "score") || hasField(value, "choice");
}

/** Normalize SDK response / mock into a name → answer mapping. */
export function answersView(raw: unknown): Record<string, unknown> {
  if (raw == null) return {};

  if (hasField(raw, "answers")) {
    const answers = attr(raw, "answers");
    if (answers != null && typeof answers === "object" && !Array.isArray(answers)) {
      return answers as Record<string, unknown>;
    }
    return {};
  }

  const combined: Record<string, unknown> = {};
  for (const bucket of ["nouls", "scores", "choices"] as const) {
    const group = attr(raw, bucket);
    if (group != null && typeof group === "object" && !Array.isArray(group)) {
      Object.assign(combined, group);
    }
  }
  if (Object.keys(combined).length > 0) return combined;

  if (typeof raw === "object" && !Array.isArray(raw)) {
    const filtered: Record<string, unknown> = {};
    for (const [name, value] of Object.entries(raw as Record<string, unknown>)) {
      if (isAnswerLike(value)) filtered[name] = value;
    }
    return filtered;
  }
  return {};
}

/** Pull named numeric signals from a TypeSafe response or mock dict. */
export function extractScores(raw: unknown): Record<string, number> {
  const scores: Record<string, number> = {};
  for (const [name, answer] of Object.entries(answersView(raw))) {
    const noul = asFloat(attr(answer, "noul"));
    if (noul != null) {
      scores[name] = noul;
      continue;
    }
    const score = asFloat(attr(answer, "score"));
    if (score != null) {
      scores[name] = score;
      continue;
    }
    const choice = attr(answer, "choice");
    const probs = attr(answer, "probabilities");
    if (choice != null && probs != null && typeof probs === "object" && !Array.isArray(probs)) {
      const p = asFloat((probs as Record<string, unknown>)[String(choice)]);
      if (p != null) scores[name] = p;
    }
  }
  return scores;
}

export interface DecideOptions {
  /** Names that must appear in the response; missing names contribute `review`. */
  expectedChecks?: Iterable<string>;
}

function pushTrigger(
  triggered: Set<Verdict>,
  triggers: Trigger[],
  verdict: Verdict,
  trigger: Trigger,
): void {
  triggered.add(verdict);
  triggers.push(trigger);
}

function lowConfidenceTrigger(
  name: string,
  confidence: number | null,
  missing: boolean,
  minConfidence: number,
): Trigger {
  const message = missing
    ? `${name} confidence missing (< min_confidence ${minConfidence})`
    : `${name} confidence=${(confidence as number).toFixed(2)} < min_confidence (${minConfidence})`;
  return { code: "low_confidence", check: name, message };
}

/**
 * Apply policy thresholds to TypeSafe answers.
 *
 * Precedence: any `block` wins over `review`; otherwise `allow`.
 * Missing expected checks, empty responses, and low/missing Score/Choice
 * confidence contribute `review` (never silent `allow`).
 */
export function decide(raw: unknown, policy: Policy, options: DecideOptions = {}): CheckResult {
  const answers = answersView(raw);
  const scores = extractScores(raw);
  const triggers: Trigger[] = [];
  const triggered = new Set<Verdict>();

  const expected = options.expectedChecks != null ? Array.from(options.expectedChecks) : undefined;
  if (expected != null) {
    for (const name of expected) {
      if (!(name in answers)) {
        pushTrigger(triggered, triggers, "review", {
          code: "missing_check",
          check: name,
          message: `${name} missing from TypeSafe response`,
        });
      }
    }
    if (expected.length === 0 && Object.keys(answers).length === 0) {
      pushTrigger(triggered, triggers, "review", {
        code: "empty_response",
        check: null,
        message: "TypeSafe response contained no answers",
      });
    }
  } else if (Object.keys(answers).length === 0) {
    pushTrigger(triggered, triggers, "review", {
      code: "empty_response",
      check: null,
      message: "TypeSafe response contained no answers",
    });
  }

  for (const [name, answer] of Object.entries(answers)) {
    if (hasField(answer, "noul")) {
      const noul = asFloat(attr(answer, "noul"));
      if (noul != null) {
        if (noul >= policy.blockThreshold && policy.blockChecks.has(name)) {
          pushTrigger(triggered, triggers, "block", {
            code: "noul_block",
            check: name,
            message: `${name}=${noul.toFixed(2)} >= block_threshold (${policy.blockThreshold})`,
          });
        } else if (noul >= policy.reviewThreshold) {
          pushTrigger(triggered, triggers, "review", {
            code: "noul_review",
            check: name,
            message: `${name}=${noul.toFixed(2)} >= review_threshold (${policy.reviewThreshold})`,
          });
        }
      }
      if (hasField(answer, "confidence")) {
        const confidence = asFloat(attr(answer, "confidence"));
        if (confidence != null && confidence < policy.minConfidence) {
          pushTrigger(
            triggered,
            triggers,
            "review",
            lowConfidenceTrigger(name, confidence, false, policy.minConfidence),
          );
        }
      }
      continue;
    }

    if (hasField(answer, "score")) {
      const score = asFloat(attr(answer, "score"));
      const confidence = asFloat(attr(answer, "confidence"));
      const missingConfidence = !hasField(answer, "confidence") || confidence == null;
      if (missingConfidence || (confidence != null && confidence < policy.minConfidence)) {
        pushTrigger(
          triggered,
          triggers,
          "review",
          lowConfidenceTrigger(name, confidence, missingConfidence, policy.minConfidence),
        );
      }
      const cutoffs = policy.scoreCutoffs(name);
      if (score != null && cutoffs != null) {
        if (score >= cutoffs.block) {
          pushTrigger(triggered, triggers, "block", {
            code: "score_block",
            check: name,
            message: `${name}=${score.toFixed(2)} >= score_block (${cutoffs.block})`,
          });
        } else if (score >= cutoffs.review) {
          pushTrigger(triggered, triggers, "review", {
            code: "score_review",
            check: name,
            message: `${name}=${score.toFixed(2)} >= score_review (${cutoffs.review})`,
          });
        }
      }
      continue;
    }

    if (hasField(answer, "choice") || hasField(answer, "confidence")) {
      const confidence = asFloat(attr(answer, "confidence"));
      const missingConfidence = !hasField(answer, "confidence") || confidence == null;
      if (missingConfidence || (confidence != null && confidence < policy.minConfidence)) {
        pushTrigger(
          triggered,
          triggers,
          "review",
          lowConfidenceTrigger(name, confidence, missingConfidence, policy.minConfidence),
        );
      }
    }
  }

  const verdict: Verdict = PRECEDENCE.find((v) => triggered.has(v)) ?? "allow";
  return {
    verdict,
    reasons: triggers.map((t) => t.message),
    scores,
    raw,
    triggers,
    codes: triggers.map(qualifiedTrigger),
  };
}
